mirrord-layer/src/tcp.rs

```diff
@@ -0,0 +1,130 @@
+use std::{
+    collections::HashSet,
+    net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr},
+};
+
+use anyhow::Result;
+/// TCP Traffic management, common code for stealing & mirroring
+use async_trait::async_trait;
+use mirrord_protocol::{NewTCPConnection, TCPClose, TCPData};
+use tokio::{net::TcpStream, sync::mpsc::Sender};
+use tracing::error;
+
+use crate::{
+    common::Listen,
+    sockets::{SocketInformation, CONNECTION_QUEUE},
+};
+
+type TrafficHandlerInputSender = Sender<TrafficHandlerInput>;
+
+#[derive(Debug)]
+pub enum TrafficHandlerInput {
+    Listen(Listen),
+    NewConnection(NewTCPConnection),
+    Data(TCPData),
+    Close(TCPClose),
+}
+
+/// To be used by traffic stealer
+// pub enum TrafficOut {}
+
+/// Struct for controlling the traffic handler struct
+pub struct TCPApi {
+    outgoing: TrafficHandlerInputSender,
+    // This is reserved for stealing API.
+    // #[allow(dead_code)]
+    // incoming: Receiver<TrafficOut>,
+}
+
+impl TCPApi {
+    pub fn new(outgoing: TrafficHandlerInputSender) -> Self {
+        Self { outgoing }
+    }
+    pub async fn send(&self, msg: TrafficHandlerInput) -> Result<()> {
+        Ok(self.outgoing.send(msg).await?)
+    }
+
+    // This is reserved for stealing API.
+    // #[allow(dead_code)]
+    // pub async fn recv(&mut self) -> Option<TrafficOut> {
+    //     self.incoming.recv().await
+    // }
+
+    pub async fn listen_request(&self, listen: Listen) -> Result<()> {
+        self.send(TrafficHandlerInput::Listen(listen)).await
+    }
+
+    pub async fn new_tcp_connection(&self, conn: NewTCPConnection) -> Result<()> {
+        self.send(TrafficHandlerInput::NewConnection(conn)).await
+    }
+
+    pub async fn tcp_data(&self, data: TCPData) -> Result<()> {
+        self.send(TrafficHandlerInput::Data(data)).await
+    }
+
+    pub async fn tcp_close(&self, close: TCPClose) -> Result<()> {
+        self.send(TrafficHandlerInput::Close(close)).await
+    }
+}
+
+#[async_trait]
+pub trait TCPHandler {
+    fn ports(&mut self) -> &HashSet<Listen>;
+    fn ports_mut(&mut self) -> &mut HashSet<Listen>;
+
+    /// Returns true to let caller know to keep running
+    async fn handle_incoming_message(&mut self, msg: Option<TrafficHandlerInput>) -> Result<bool>
```

Comment:
Function signature looks a bit weird to me:

1. Why not take the `Traffic<_>` here, instead of `Option<Traffic<_>>`?;
2. Why `Result<bool>` and not `Result<()>` and check `result.is_ok()` on the caller side?

Reply:
Didn't understand 1 btw
We need to keep it running even after an error. Any suggestion how to mark error ignorable?

Reply:
1. Leave the function caller responsible for checking if it's an option or not, and then it passes down just the thing, instead of `Option` thing.

Make a `LayerError::KeepRunning` or something like that, and check for it.

Reply:
`LayerError::KeepRunning` would be 1 error but if we have many different errors that we wish to ignore, you have a suggestion for a nicer approach?

Reply:
`map_err` for all ignorable errors? Not sure.